judge/utils/tests/test_file_cache.py

```diff
@@ -0,0 +1,111 @@
+import gzip
+import os
+import shutil
+import tempfile
+from io import UnsupportedOperation
+from urllib.parse import urljoin
+
+from django.test import SimpleTestCase
+
+from judge.utils.file_cache import HashFileCache
+
+
+class FileCacheTestCase(SimpleTestCase):
+    def setUp(self):
+        self.temp_dir = tempfile.mkdtemp()
+        self.cache = HashFileCache(self.temp_dir, 'https://example.com')
+        self.gzip_temp_dir = tempfile.mkdtemp()
+        self.gzip_cache = HashFileCache(self.gzip_temp_dir, 'https://example.com', gzip=True)
+
+        # This acts as a test for create() as well as setup for other tests
+        self.cache.create('foo')
+        self.cache.create('foo')
+        self.gzip_cache.create('foo')
+        self.gzip_cache.create('foo')
+
+    def test_init(self):
+        root = '/foo/bar/'
+        url = 'https://example.com'
+
+        # Test default
+        cache = HashFileCache(root, url)
+        self.assertEqual(cache.root, root)
+        self.assertEqual(cache.url, url)
+        self.assertFalse(cache.gzip)
+
+        # Test setting gzip
+        cache = HashFileCache(root, url, gzip=True)
+        self.assertEqual(cache.root, root)
+        self.assertEqual(cache.url, url)
+        self.assertTrue(cache.gzip)
+
+    def test_create_error(self):
+        with self.assertRaises(NotADirectoryError):
+            cache = HashFileCache('/', '')
+            cache.create('/dev/null/reee')
+
+    def test_has_file(self):
+        self.assertFalse(self.cache.has_file('random', 'bar'))
+        with tempfile.NamedTemporaryFile(dir=os.path.join(self.temp_dir, 'foo')) as f:
+            fname = f.name
+            self.assertTrue(self.cache.has_file('foo', fname))
+        self.assertFalse(self.cache.has_file('foo', fname))
+
+    def test_get_path(self):
+        self.assertTrue(self.cache.get_path('foo', 'bar'), os.path.join(self.temp_dir, 'foo', 'bar'))
+        self.assertTrue(self.cache.get_path('bar', 'foo'), os.path.join(self.temp_dir, 'bar', 'foo'))
+
+    def test_get_url(self):
+        urls = [
+            ('foo', 'bar'),
+            ('foo/', 'bar/'),
+            ('/foo', '/bar'),
+            ('foo/', '/bar'),
+            ('/foo', 'bar/'),
+            ('/foo', '/bar/'),
+        ]
+        for hash, file in urls:
+            self.assertTrue(self.cache.get_url(hash, file), urljoin(self.cache.url, '%s/%s' % (hash, file)))
+
+    def test_read_file(self):
+        # The directory removal in tear down will clean up this file, so we don't have to cleanup here
+        with tempfile.NamedTemporaryFile(dir=os.path.join(self.temp_dir, 'foo'), delete=False) as f:
```

Comment:
Do we really need to use `NamedTemporaryFile` when we can just create a file with a fixed name?

Reply:
I suppose we could, but it would have to be unique across all tests since there's an option to run tests in parallel.

Reply:
`setUp` and `tearDown` are run once per test case, so that's not a problem.

Reply:
Wouldn't there be a possible race conditions?
---
On another note, in basically all my test cases, I don't need the `setUp` and `tearDown` to run for every test. (In fact I didn't realize it happened until quite late.) This would also very very marginally increase performance.

Reply:
It's not a problem because `self.temp_dir` is different for every test case. I wouldn't recommend (accidentally or not) depending on the test execution order in any particular group of test cases, so it's better to setup and teardown every case.

Reply:
Oh right, it's not an issue for the very reason that `setUp` and `tearDown` happen every test.
> I wouldn't recommend (accidentally or not) depending on the test execution order in any particular group of test cases, so it's better to setup and teardown every case.

Agreed.